### DISCUSSION TITLE
[MTR05] week5 LIOJ 1018

### DISCUSSION THREAD

**Discussion by YO-0115** [Lidemy/forum#103] [程式導師實驗計畫]
這題我是依照老師在 `1035簡易排序` 中所教的概念下去寫的。
一個籃子有幾顆球，藉此算出每層有多長。

在自己的電腦上測試是對的，但就是無法 AC，
也有在各個地方加 log 看過了，但看不出來是哪邊錯了。
想請問哪裡是我沒注意到的地方嗎?  謝謝
```
Input

第一行是一個正整數 N，1 <= N <= 1000，代表一共有幾個數字

第二行會有 N 個以空白分隔的數字 Ai，1 <= Ai ​<= 1000，已經由小到大排序好，代表每一個階層的長度
```
```
Output

請你求出大平台（也就是長度最長的平台）的長度為何
```
```
Sample Input 1

10
1 1 2 2 2 3 3 3 4 4
```
```
Sample Output 1

3
```

``` js
function solve(lines) {
  let temp = lines[1].split(' ')
  

  let arr = []
  for (let i=1; i<=lines[0]; i++) {
    arr[i] = 0
  }
  

  for (let i=0; i<temp.length; i++) {
    let n = Number(temp[i])
    arr[n]++
  }
  

  let longest = 0
  for (let i=1; i<arr.length; i++) {
    if (arr[i] > longest) {
      longest = arr[i]
    }
  }
  console.log(longest)
}
```

**Comment by nuoxoxo** [comment]
大膽猜測：
你一開始準備用來做 count 的 arr ，他的 arr.length 假定了 lines[0] 中每個元素只出現一次，然後安排好足夠的位置

例如，如果 input 是
```
10
1 1 2 3 4 5 6 7 7 7 8 8 8 8
```
會 log 出 8 的長度 4
而不會是 3

**Reply by YO-0115** [reply]
但如果是上述的 input ，答案應該是 4 沒錯吧 ?

**Reply by nuoxoxo** [reply]
> 但如果是上述的 input ，答案應該是 4 沒錯吧 ?

但 lines[0] 規定了是 10 ，所以不能數到 `777` 後面的 `8888` 那裡

**Reply by YO-0115** [reply]
剛剛又看了一下題目，lines[0] 是 10 的話，那 lines[1]，也只會有 10 個數字。
應該是這樣吧?

**Reply by nuoxoxo** [reply]
> 剛剛又看了一下題目，lines[0] 是 10 的話，那 lines[1]，也只會有 10 個數字。
> 應該是這樣吧?

是的。壞消息是，我把 code 不斷整理不斷 submit 都得到 WA 。所以祝你好運。我直覺是這邊不用 counting sort 可能比較好。我的做法是宣告一個 `counter = 1`，如果 i 元素和 i + 1 元素是相等的話就 `counter++` ，否則平台就斷了，這時就把 counter 接住：`longest = counter; counter = 1` 。最後 longest 就是答案

**Reply by YO-0115** [reply]
原來，我得要時間來理解一下。
謝謝你。

**Reply by cwc329** [reply]
試試看這組
```
3
11 11 12
```

**Reply by cmtilo** [reply]
好

**Reply by allenliaooooooo** [reply]
題目應該是不會陰人吧...XD
第一行給 10 就一定會是 10 個數字

**Reply by EshauHuang** [reply]
你的 n 是 arr 矩陣的 index 值，但你的 index 只有 1-10，所以如果你要顯示正確的值你應該把 index 1-1000(題目最大數字)全部給 0 這個值，不然 index 值 10 以上一定跑不出來。

![Imgur](https://i.imgur.com/U0fmUyo.png)

這個圖片是帶入
``` 
3
11 11 12
```
 arr=0 及 arr++ 後的值

將
```
for (let i=1; i<=lines[0]; i++) {
    arr[i] = 0
  }
```
改成這樣
```
  for (let i=1; i<=1000; i++) {
    arr[i] = 0
  }
```
測試過可以順利 AC

**Reply by YO-0115** [reply]
原來如此，真的 AC 了。
我一開始以為，平台只會從 1 樓開始。
謝謝各位的解惑。
意外收穫 2 種解法😁

**Reply by cmtilo** [reply]
我也犯了同樣的錯誤~~以為萬丈高樓平地起...

**Reply by EshauHuang** [reply]
不過你 index 0 要注意還是沒有值喔。

**Reply by YO-0115** [reply]
想請問一下，以這題來說，已經知道不可能會有 0 的這個長度，那還是要給它一個籃子嗎 ?
還是讓它沒有值也沒關係呢 ??

**Reply by EshauHuang** [reply]
> 想請問一下，以這題來說，已經知道不可能會有 0 的這個長度，那還是要給它一個籃子嗎 ?
> 
> 還是讓它沒有值也沒關係呢 ??

依題目的話確實不加也不會有影響。但矩陣這樣空著沒有值感覺就是有點問題，這邊我不太確定你可能要再問問其他人。

**Reply by YO-0115** [reply]
好的，謝謝你

**Reply by cwc329** [reply]
也不一定要用 array，
可以用 object 記錄，
key 就是數字，
value 就是出現次數。
```js
const obj = {
  3: 1
  8: 7
  10: 5
};
```
